fraud/bad_encoding.go

```diff
@@ -0,0 +1,179 @@
+package fraud
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/tendermint/tendermint/pkg/consts"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// Shares contains all shares from row/col
+	// Shares that did not pass verification in rmst2d will be nil
+	// For non-nil shares MerkleProofs are computed
+	Shares []*Share
+	// Position represents the number of row/col where ErrByzantineRow/ErrByzantineColl occurred
+	Position uint8
+	isRow    bool
+}
+
+func CreateBadEncodingFraudProof(
+	height uint64,
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	errShares [][]byte,
+) (Proof, error) {
+	var tree ErasuredNamespacedMerkleTree
+	shares := make([]*Share, len(errShares))
+
+	for index, share := range errShares {
+		if share != nil {
+			tree = NewErasuredNamespacedMerkleTree(uint64(len(errShares) / 2))
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+			if err := buildTreeFromDataRoot(&tree, data, index); err != nil {
```

Comment:
Q: Why is this called `*FromDataRoot`? From what I understand it takes the row or column data (and not a root)? Or am I missing sth obvious?

Reply:
Also, I think it might be better if the buildTree method constructs the tree entirely, i.e. put the `tree = NewErasuredNamespacedMerkleTree(uint64(len(errShares) / 2))` into the buildTree method and return a pointer instead.

Reply:
Reworked.